server/apis/v1/handler.go

```diff
@@ -46,282 +68,831 @@ type handler struct {
 		return
 	}
-	c.JSON(http.StatusOK, vertices.Items)
+
+	oldSpec, err := h.numaflowClient.Pipelines(ns).Get(context.Background(), pipeline, metav1.GetOptions{})
+	if err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to fetch pipeline %q namespace %q, %s", pipeline, ns, err.Error()))
+		return
+	}
+
+	updatedSpec, ok := reqBody.(*dfv1.Pipeline)
+	if !ok {
+		h.respondWithError(c, "Failed to convert request body to pipeline spec")
+		return
+	}
+	// Validate the namespace of the request
+	err = validateNamespace(h, updatedSpec, ns)
+	if err != nil {
+		h.respondWithError(c, err.Error())
+		return
+	}
+
+	// pipeline name in the URL should be same as spec name
+	if pipeline != updatedSpec.Name {
+		h.respondWithError(c, fmt.Sprintf("pipeline name %q is immutable", pipeline))
+		return
+	}
+
+	oldSpec.Spec = updatedSpec.Spec
+	updatedSpec.Namespace = ns
+	isValid := validatePipelineSpec(h, oldSpec, updatedSpec, ValidTypeUpdate)
+	if isValid != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to update pipeline %q, %s", pipeline, isValid.Error()))
+		return
+	}
+	// If Validation flag is set to true, return without updating the pipeline
+	if dryRun == "true" {
+		c.JSON(http.StatusOK, NewNumaflowAPIResponse(nil, nil))
+		return
+	}
+
+	if _, err := h.numaflowClient.Pipelines(ns).Update(context.Background(), oldSpec, metav1.UpdateOptions{}); err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to update pipeline %q, %s", pipeline, err.Error()))
+		return
+	}
+
+	c.JSON(http.StatusOK, NewNumaflowAPIResponse(nil, nil))
 }
 
-// GetVertex is used to provide the vertex spec
-func (h *handler) GetVertex(c *gin.Context) {
-	vertices, err := h.numaflowClient.Vertices(c.Param("namespace")).List(context.Background(), metav1.ListOptions{
-		LabelSelector: fmt.Sprintf("%s=%s,%s=%s", dfv1.KeyPipelineName, c.Param("pipeline"), dfv1.KeyVertexName, c.Param("vertex")),
-	})
+// DeletePipeline is used to delete a given pipeline
+func (h *handler) DeletePipeline(c *gin.Context) {
+	ns, pipeline := c.Param("namespace"), c.Param("pipeline")
+
+	if err := h.numaflowClient.Pipelines(ns).Delete(context.Background(), pipeline, metav1.DeleteOptions{}); err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to delete pipeline %q, %s", pipeline, err.Error()))
+		return
+	}
+
+	c.JSON(http.StatusOK, NewNumaflowAPIResponse(nil, nil))
+}
+
+// PatchPipeline is used to patch the pipeline spec to achieve operations such as "pause" and "resume"
+func (h *handler) PatchPipeline(c *gin.Context) {
+	ns, pipeline := c.Param("namespace"), c.Param("pipeline")
+
+	patchSpec, err := io.ReadAll(c.Request.Body)
 	if err != nil {
-		c.JSON(http.StatusInternalServerError, err.Error())
+		h.respondWithError(c, fmt.Sprintf("Failed to parse request body, %s", err.Error()))
 		return
 	}
-	if len(vertices.Items) == 0 {
-		c.JSON(http.StatusNotFound, fmt.Sprintf("Vertex %q not found", c.Param("vertex")))
+
```

Comment:
Add a `TODO`, we should validate the patched data as well, e.g. only allow lifecycle to be patched.